storage/reports: convert some fields to timestamptz

andreimatei opened the pull request:
This patch adds a migration converting the columns
system.reports_meta.generated and system.replication_straint_stats from
timestamp to timestamptz.

Timestamp is a fairly useless data type with surprising behavior -
clients generally interpret a timestamp value to be in their local time
zone. For representing a point in time, one wants a timestamptz.

Release note (sql change): The system.reports_meta.generated and
system.replication_straint_stats columns now have type timestamptz
instead of timestamp.

andreimatei committed:
storage/reports: convert some fields to timestamptz

This patch adds a migration converting the columns
system.reports_meta.generated and system.replication_straint_stats from
timestamp to timestamptz.

Timestamp is a fairly useless data type with surprising behavior -
clients generally interpret a timestamp value to be in their local time
zone. For representing a point in time, one wants a timestamptz.

Release note (sql change): The system.reports_meta.generated and
system.replication_straint_stats columns now have type timestamptz
instead of timestamp.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41905)
<!-- Reviewable:end -->

andreimatei commented:
i want to backport this

knz commented on pkg/storage/reports/constraint_stats_report.go:
```diff
@@ -197,9 +197,9 @@ func (r *replicationConstraintStatsReportSaver) loadPreviousVersion(
 
 		// if the row is nil then this is the first time we are running and the reload is needed.
 		if row != nil {
```
probably `if len(row) > 0` here instead. Otherwise if you have a non-nil but empty slice, the case below will panic.

knz commented on pkg/storage/reports/critical_localities_report.go:
```diff
@@ -110,7 +110,7 @@ func (r *replicationCriticalLocalitiesReportSaver) loadPreviousVersion(
-			generated, ok := row[0].(*tree.DTimestamp)
+			generated, ok := row[0].(*tree.DTimestampTZ)
 			if !ok {
 				return errors.Errorf("Expected to get time from system.reports_meta but got %+v", row)
```
if you change the error above to use `row[0]` probably you want the same here.

andreimatei replied:
done

knz commented on pkg/storage/reports/critical_localities_report.go:
```diff
@@ -110,7 +110,7 @@ func (r *replicationCriticalLocalitiesReportSaver) loadPreviousVersion(
 
 		// if the row is nil then this is the first time we are running and the reload is needed.
 		if row != nil {
```
ditto

andreimatei replied:
done

knz commented on pkg/storage/reports/replication_stats_report.go:
```diff
@@ -113,9 +113,9 @@ func (r *replicationStatsReportSaver) loadPreviousVersion(
 
 		// if the row is nil then this is the first time we are running and the reload is needed.
 		if row != nil {
```
ditto

andreimatei replied:
done

knz reviewed (approved):
LGTM with nits

darinpp reviewed (approved):
Looks good.

andreimatei commented:
TFTR

andreimatei commented:
I've added another preamble commit. Without it, the contents of the `system.event_log` are pretty non-deterministic after migrations like the one I've added, which makes some logic tests unhappy.
I'd backport that too, but if y'all don't think so I can also paper over the problem in the backport.

knz reviewed (commented):
> This patch improves the mechanism such that migrations included in the
> bootstrap image are never run.

Please extend that commit sentence with a clarification of when it does not run.

Also as discussed separately I encourage you to split that commit in a separate PR, and seek an additional review by Tobias.

> Reviewed 6 of 6 files at r2, 5 of 5 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41905#-:-LrzmWKe9wMMjg9_jbAH:b-g6dds7)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andreimatei commented:
Moved the migration stuff to https://github.com/cockroachdb/cockroach/pull/41914. Hacked my way around the event_log test here.

knz reviewed (approved):
LGTM if tests are happy.

> Reviewed 5 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41905#-:-LrzxmB29aO3mB5pVExo:b-i310zv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->
